net/xia/ppal_ether/main.c

```diff
@@ -191,10 +502,65 @@ static struct ether_interface *eint_init(struct net_device *dev)
 	return eint;
 }
 
+/* Caller must hold RTNL lock, and makes sure that nobody adds entries
+ * in eint->list_interface_common_addr while it's running.
+ */
+static void free_neighs_by_interface(struct ether_interface *eint)
+{
+	struct net_device *dev;
+	struct net *net;
+	struct xip_ppal_ctx *ctx;
+	struct fib_xid_table *xtbl;
+
+	ASSERT_RTNL();
+
+	dev = eint->dev;
+	net = dev_net(dev);
+	ctx = xip_find_my_ppal_ctx_vxt(net, ether_vxt);
+	xtbl = ctx->xpc_xtbl;
+
+	while (1) {
+		struct fib_xid_ether_main *mether;
+		u8 xid[XIA_XID_MAX];
+		struct fib_xid *fxid;
+		u32 bucket;
+
+		/* Obtain xid of the first entry in
+		 * @eint->list_interface_common_addr.
+		 * We use rcu_read_lock() here to allow one to remove
+		 * entries in parallel.
+		 */
+		rcu_read_lock();
+		mether = list_first_or_null_rcu(&eint->
+						list_interface_common_addr,
+					    struct fib_xid_ether_main,
+					    interface_common_addr);
+		if (!mether) {
+			rcu_read_unlock();
+			break;
+		}
+		memmove(xid, mether->xem_common.fx_xid, XIA_XID_MAX);
+		rcu_read_unlock();
+
+		/* We don't lock eint->interface_lock to avoid deadlock. */
+		fxid = ether_rt_iops->fxid_find_lock(&bucket, xtbl, xid);
+		if (fxid) {
+			BUG_ON(fxid->fx_table_id != XRTABLE_MAIN_INDEX);
+			detach_neigh_to_interface(mether);
```

Comment:
The line `detach_neigh_to_interface(mether);` must go to `main_free_ether()` in order to avoid upseting RCU readers.